android/src/main/java/com/amplitude/android/utilities/Session.kt

```diff
@@ -0,0 +1,99 @@
+package com.amplitude.android.utilities
+
+import com.amplitude.android.Amplitude
+import com.amplitude.android.Configuration
+import com.amplitude.common.Logger
+import com.amplitude.core.State
+import com.amplitude.core.Storage
+import com.amplitude.core.events.BaseEvent
+import java.util.concurrent.atomic.AtomicLong
+
+class Session(
+    private var configuration: Configuration,
+    private var storage: Storage? = null,
+    private var state: State? = null,
+) {
+    private val _sessionId = AtomicLong(configuration.sessionId ?: -1L)
+
+    val sessionId: Long
+        get() {
+            return _sessionId.get()
+        }
+
+    internal var lastEventId: Long = 0
+    var lastEventTime: Long = -1L
+
+    init {
+        val currentSessionId = configuration.sessionId
+            ?: (storage?.read(Storage.Constants.PREVIOUS_SESSION_ID)?.toLongOrNull() ?: -1)
+
+        _sessionId.set(currentSessionId)
+        state?.sessionId = currentSessionId
+        lastEventId = storage?.read(Storage.Constants.LAST_EVENT_ID)?.toLongOrNull() ?: 0
+        lastEventTime = storage?.read(Storage.Constants.LAST_EVENT_TIME)?.toLongOrNull() ?: -1
+    }
+
+    suspend fun startNewSessionIfNeeded(timestamp: Long): Iterable<BaseEvent>? {
+        if (inSession() && isWithinMinTimeBetweenSessions(timestamp)) {
+            refreshSessionTime(timestamp)
+            return null
+        }
+        return startNewSession(timestamp)
+    }
+
+    suspend fun setSessionId(timestamp: Long) {
+        _sessionId.set(timestamp)
+        storage?.write(Storage.Constants.PREVIOUS_SESSION_ID, timestamp.toString())
+        state?.sessionId = timestamp
+    }
+
+    suspend fun startNewSession(timestamp: Long): Iterable<BaseEvent> {
+        val sessionEvents = mutableListOf<BaseEvent>()
+        // If any trackingSessionEvents is false (default value is true), means it is manually set
+        @Suppress("DEPRECATION")
+        val trackingSessionEvents = configuration.trackingSessionEvents && configuration.defaultTracking.sessions
+
+        // end previous session
+        if (trackingSessionEvents && inSession()) {
+            val sessionEndEvent = BaseEvent()
+            sessionEndEvent.eventType = Amplitude.END_SESSION_EVENT
+            sessionEndEvent.timestamp = if (lastEventTime > 0) lastEventTime else null
+            sessionEndEvent.sessionId = sessionId
+            sessionEvents.add(sessionEndEvent)
+        }
+
+        // start new session
+        setSessionId(timestamp)
+        refreshSessionTime(timestamp)
+        if (trackingSessionEvents) {
```

Comment:
How can we fire the identify event before the start session?

Reply:
Hmm, not sure yet. I am mostly focused on getting session id working for session replay. Will leave identify fix for another PR.

Reply:
I think at some point we'll need to change the code so that firing of events happens outside session management. 

The logic shoudl be something like 

```
if (session.hasSessionExpired()) {
    fireEndSessionEvent();
    session.createNewSession();
    fireAnyPreSessionInit();
    fireStartSessionEvent();
}
```

Reply:
Not a blocker for this PR